core/src/Streamly/Internal/FileSystem/PosixPath.hs

```diff
@@ -0,0 +1,362 @@
+-- not validate that yet when parsing a path.
+-- Path components may have limits.
+-- Total path length may have a limit.
+
+-- | A type representing file system paths on OS_NAME.
+--
+-- A OS_PATH is validated before construction unless unsafe constructors are used
+-- to create it. Rules and invariants maintained by the safe construction
+-- methods are as follows:
+--
+-- * Does not contain a null character.
+-- * Does not have a trailing separator except in the root path.
+-- * Does not have a trailing @.@ component.
+-- * Does not have consecutive separators except in UNC prefixes on Windows.
+-- * Does not contain @\/.\/@ path components except in a UNC prefix on
+--   Windows.
+--
+-- Note that in some cases the file system may perform unicode normalization on
+-- paths (e.g. Apple HFS), it may cause surprising results as the path used by
+-- the user may not have the same bytes as later returned by the file system.
+newtype OS_PATH = OS_PATH (Array WORD_TYPE)
+
+-- Show instance prints raw bytes without any decoding for rountdtripping. We
+-- can use a Lax Utf8 decoding and print it as a string for convenience? Should
+-- we print raw path as a string instead, may be useful for ascii chars but
+-- utf8 encoded chars may be unprintable.  Better use toString if you want to
+-- pretty print the path.
+{-
+instance Show OS_PATH where
+    show (OS_PATH x) = show x
+-}
+
+-- XXX The Eq instance needs to make sure that the paths are equivalent. If we
+-- normalize the paths we can do a byte comparison. However, on windows paths
+-- are case insensitive but the case is preserved, therefore, we cannot
+-- normalize and need to do case insensitive comparison.
+
+instance IsPath OS_PATH OS_PATH where
+    unsafeFromPath = id
+    fromPath = pure
+    toPath = id
+
+-- XXX Use rewrite rules to eliminate intermediate conversions for better
+-- efficiency. If the argument path is already verfied for a property, we
+-- should not verify it again e.g. if we adapt (Loc path) as (Loc (Dir path))
+-- then we should not verify it to be Loc again.
+
+-- XXX castPath?
+
+-- | Convert a path type to another path type. This operation may fail with a
+-- 'PathException' when converting a less restrictive path type to a more
+-- restrictive one. This can be used to upgrade or downgrade type safety.
+adapt :: (MonadThrow m, IsPath OS_PATH a, IsPath OS_PATH b) => a -> m b
+adapt p = fromPath (toPath p :: OS_PATH)
+
+------------------------------------------------------------------------------
+-- Path parsing utilities
+------------------------------------------------------------------------------
+
+-- XXX rather have a dropEmptySegments?
+
+-- | If the path is @//@ the result is @/@. If it is @a//@ then the result is
+-- @a@.
+{-# INLINE dropTrailingSeparators #-}
+dropTrailingSeparators :: OS_PATH -> OS_PATH
+dropTrailingSeparators (OS_PATH arr) =
+    OS_PATH (Common.dropTrailingSeparators Common.Posix arr)
```

Comment:
You have `Common.Posix` here even when `IS_WINDOWS` is defined. Will this work?

Reply:
I'm guessing this should be `Common.OS_NAME`

Reply:
This is not right, should be fixed.